### PR TITLE
Document how upstream OAuth 2.0 providers should be setup

### LINKED ISSUE #1940: The upstream OAuth 2.0 configuration is not documented
A lot can be inferred from the config JSON schema, but the documentation still needs to guide through configuration of upstream IDPs.
It needs:

 - documentation of the config options
 - examples with popular IDPs (Google, Dex, Keycloak, ?)

### PR DESCRIPTION
**Pull request by sandhose**
Fixes #1940

External feedback is very much welcome, preview: 

 - https://quenting-upstream-oauth-docs.matrix-authentication-service-docs.pages.dev/setup/sso
 - https://quenting-upstream-oauth-docs.matrix-authentication-service-docs.pages.dev/usage/configuration#upstream_oauth2

### CODE REVIEW

File: docs/setup/sso.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+A Jinja2 template is used as mapping for each attribute. The template currently has one `user` variable, which is an object with the claims got through the `id_token` given by the provider.
+The following default templates are used:
+
+ - `localpart`: `{{ user.preferred_username }}`
+ - `display_name`: `{{ user.name }}`
+ - `email`: `{{ user.email }}`
+
+## Multiple providers behaviour
+
+Mutiple authentication methods can be configured at the same time, in which case the authentication service will let the user choose which one to use.
```
Comment (wrjlewis) [inline docs/setup/sso.md:60]: s/Mutiple/Multiple

File: docs/setup/sso.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+      issuer: "https://<keycloak>/realms/<realm>" # TO BE FILLED
+      token_endpoint_auth_method: client_secret_basic
+      client_id: "matrix-authentication-service"
+      client_secret: "<client-secret>" # TO BE FILLED
+      scope: "openid profile email"
+      claims_imports:
+        localpart:
+          action: require
+          template: "{{ user.preferred_username }}"
+        displayname:
```
Comment (wrjlewis) [inline docs/setup/sso.md:255]: In some sections it's called `displayname`, in others it's called `display_name`. Should we be consistent with the naming?

File: docs/setup/sso.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+          action: suggest
+          template: "{{ user.name }}"
+        email:
+          action: suggest
+          template: "{{ user.email }}"
+          set_email_verification: always
+```
+
+
+## GitLab
```
Comment (wrjlewis) [inline docs/setup/sso.md:147]: s/## GitLab/### GitLab/

File: docs/setup/sso.md [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+      human_name: Authentik
+      issuer: "https://<authentik-domain>/application/o/<app-slug>/" # TO BE FILLED
+      client_id: "<client-id>" # TO BE FILLED
+      client_secret: "<client-secret>" # TO BE FILLED
+      scope: "openid profile email"
+      claims_imports:
+        localpart:
+          action: require
+          template: "{{ user.preferred_username }}"
+        display_name:
```
Comment (wrjlewis) [inline docs/setup/sso.md:96]: Sometimes this is display_name and in other sections it is displayname

File: docs/usage/configuration.md [hunk 5]
```diff
@@ -78,8 +78,8 @@ http:
 
 The following additional resources are available, although it is recommended to serve them on a separate listener, not exposed to the public internet:
 
- - `name: prometheus`: serves the a Prometheus-compatible metrics endpoint on `/metrics`, if the Prometheus exporter is enabled in `telemetry.metrics.exporter`.
```
Comment (wrjlewis) [inline docs/usage/configuration.md:81]: s/serves the a Prometheus-compatible/serves a Prometheus-compatible

(remove 'the')

**Review by wrjlewis — COMMENTED** [review]
Thanks for documenting, it's great. Just nitpicks from me on minor things.

File: docs/setup/sso.md (hunk 3 above)
Comment (sandhose) [reply docs/setup/sso.md:147]: fyi if you hit the "Add a suggestion" button in the GitHub markdown editor, it formats the change like this:

```suggestion
### GitLab
```

which I can then one-click commit with you as co-author

File: docs/setup/sso.md (hunk 2 above)
Comment (sandhose) [reply docs/setup/sso.md:255]: Good catch, should be `displayname` everywhere

File: docs/setup/sso.md [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+          template: "{{ user.email }}"
+          set_email_verification: always
+```
+
+
+### GitLab
+
+1. Create a [new application](https://gitlab.com/profile/applications).
+2. Add the `openid` scope. Optionally add the `profile` and `email` scope if you want to import the user's name and email.
+3. Add this Callback URL: `https://<auth-service-domain>/upstream/callback/<id>`
```
Comment (wrjlewis) [inline docs/setup/sso.md:151]:
```suggestion
3. Add this Callback URL: `https://<auth-service-domain>/upstream/callback/<id>`.
```
Making a simple change (and to be fair other items in the list have '.' at the end) to test what you said with suggesting changes..

Comment (wrjlewis) [reply docs/setup/sso.md:151]: Thanks for the tip

**Pull request merged**